sirepo/pkcli/test_http.py

```diff
@@ -322,60 +288,230 @@ def _uri(self, uri):
+            ),
+            self,
+        )
+        if error:
+            await c
+            return
+        with self._set_waiting_on_status(frame_before_caller=True):
+            await c
+
+    async def _get_sim_frame(self, next_request):
+        g = self._app.sim_data.frame_id(self._data, next_request, self._report, 0)
+        with self._set_waiting_on_status():
+            f = await self._app.client.get('/simulation-frame/' + g, self)
+        assert 'title' in f, \
+            '{} no title in frame={}'.format(self.pkdebug_str(), f)
+        with self._set_waiting_on_status():
+            await self._app.client.get(
+                '/download-data-file/{}/{}/{}/{}'.format(
+                    self._app.sim_type,
+                    self._sid,
+                    self._report,
+                    0,
+                ),
+                self,
+            )
+        return g
+
+    async def _run_sim(self):
+        r = self._report
+        if 'animation' in self._report.lower() and self._app.sim_type != 'srw':
+            r = 'animation'
+        return await self._app.client.post(
+            '/run-simulation',
+            PKDict(
+                # works for sequential simulations, too
+                forceRun=True,
+                models=self._data.models,
+                report=r,
+                simulationId=self._sid,
+                simulationType=self._app.sim_type,
+            ),
+            self,
+            )
+
+    async def _run_sim_until_completion(self):
+        c = True
+        e = False
+        try:
+            with self._set_waiting_on_status():
+                r = await self._run_sim()
+            t = random.randrange(cfg.run_min_secs, cfg.run_max_secs)
+            for _ in range(t):
+                if r.state == 'completed' or r.state == 'error':
+                    c = False
+                    assert r.state != 'error', \
+                        '{} unexpected error state {}'.format(
+                            self.pkdebug_str(),
+                            r,
+                        )
+                    break
+                assert 'nextRequest' in r, \
+                    '{} expected "nextRequest" in response={}'.format(
+                        self.pkdebug_str(),
+                        r,
+                    )
+                with self._set_waiting_on_status():
+                    r = await self._app.client.post('/run-status', r.nextRequest, self)
+                with self._set_waiting_on_status():
+                    await tornado.gen.sleep(1)
+            else:
+                pkdlog('{} timeout={}', self, t)
+            return r
+        except Exception:
+            e = True
+            raise
+        finally:
+            if c:
+                await self._cancel(e)
+
 def _init():
     global cfg
     if cfg:
         return
     cfg = pkconfig.init(
         server_uri=('http://127.0.0.1:8000', str, 'where to send requests'),
```

Comment:
Can I just import pkcli.service and [read its config](https://github.com/radiasoft/sirepo/blob/c71be82eff79dd2f961b00f73801a52b2e315e55/sirepo/pkcli/service.py#L180)?

Reply:
A couple things:

* the function needs to be public so pkcli._commands needs to ignore functions that have a particular prefix, e.g. internal_.
* the cfg should be exported via a function, e.g. internal_cfg(), which is called internally to handle initialization.
* Move `import sirepo.srdb` to `_run_dir`, it will limit imports.

Then I think it would be ok to import pkcli.service into test_http.

Reply:
Can I skip the first two bullet points and just put the test_http.cfg globally on the module? AFAIK, putting it on the module would do everything `_init()` currently does.

For my own knowledge, what is the advantage of moving the `import sirepo.srdb` into `_run_dir()`? I understand it's advantage in tests where we don't want global imports confusing aspects of the state but I'm unsure of the use case here.

Reply:
I don't think I was clear, and the underscores turned into italics.

It is a non-explicit coupling to import the module to init cfg. Some modules need to init later. Rather, there should be function in pkcli.service that gets called to return the cfg. We do this explicitly in feature_config. I think we should do it in other places when we share config this way.

The pkcli.service function is not a command. It technically could be, because it is not doing anything so we could skip the first bullet. If we exported service.cfg() to get the cfg, and also used that same function in the module to refer to the cfg, it would work fine.

Adopting this same model for test_http would work, too. Instead of `cfg` being a value, it is a function, and is `_cfg` is the value, just like `feature_cfg.cfg`. 

There is no "use case" for moving import srdb into `_run_dir`. It's about limiting imports and ensuring initialization order. srdb is only needed in _run_dir so we are reducing the import surface and ensuring initialization order isn't mucked with.

Reply:
Ah, make sense now. Thanks for the detailed explanation.